corehq/util/urlsanitize/test/test_urlsanitize.py

```diff
@@ -1,54 +1,65 @@
+        ('http://Localhost/', (RAISE, PossibleSSRFAttempt('is_loopback'))),
+        ('http://169.254.169.254/latest/meta-data', (RAISE, PossibleSSRFAttempt('is_link_local'))),
+        ('http://2852039166/', (RAISE, PossibleSSRFAttempt('is_link_local'))),
+        ('http://7147006462/', (RAISE, PossibleSSRFAttempt('is_link_local'))),
+        ('http://0xA9.0xFE.0xA9.0xFE/', (RAISE, PossibleSSRFAttempt('is_link_local'))),
+        ('http://0x41414141A9FEA9FE/', (RAISE, PossibleSSRFAttempt('is_link_local'))),
+        ('http://0xA9FEA9FE/', (RAISE, PossibleSSRFAttempt('is_link_local'))),
+        ('http://0251.0376.0251.0376/', (RAISE, PossibleSSRFAttempt('is_link_local'))),
+        ('http://0251.00376.000251.0000376/', (RAISE, PossibleSSRFAttempt('is_link_local'))),
+        ('http://169.254.169.254.xip.io/', (RAISE, PossibleSSRFAttempt('is_link_local'))),
+        ('http://10.124.10.11', (RAISE, PossibleSSRFAttempt('is_private'))),
+        ('some-non-url', (RAISE, InvalidURL())),
+    ]
 
 
 def test_example_suite():
     for input_url, (result, value) in SUITE:
         print(input_url)
+
         if result is RETURN:
-            assert sanitize_user_input_url(input_url) == value, \
-                f"sanitize_url({input_url!r}) should be {value!r} was {sanitize_user_input_url(input_url)}"
+            eq(sanitize_user_input_url(input_url), value)
         elif result is RAISE:
-            try:
-                sanitize_user_input_url(input_url)
-            except value.__class__ as e:
-                if value.__class__ == PossibleSSRFAttempt:
-                    assert e.reason == value.reason, \
-                        f"sanitize_url({input_url!r} should raise {value.__class__.__name__}({value.reason}) raised {value.__class__.__name__}({e.reason})"
-                else:
-                    assert str(e) == str(value), \
-                        f"sanitize_url({input_url!r} should raise {value.__class__.__name__}({str(value)}) raised {value.__class__.__name__}({str(e)})"
+            if type(value) == PossibleSSRFAttempt:
+                with assert_raises(PossibleSSRFAttempt, msg=lambda e: eq(e.reason, value.reason)):
+                    sanitize_user_input_url(input_url)
+            else:
+                with assert_raises(type(value), msg=str(value)):
+                    sanitize_user_input_url(input_url)
         else:
             raise Exception("result in suite should be RETURN or RAISE")
 
 
-if __name__ == '__main__':
-    test_example_suite()
+def test_rebinding():
+    """
+    this test doesn't do much, it just checks that a known rebinding endpoint
+    and checks the output is one of the two expected values
+    """
+    url = 'http://A.8.8.8.8.1time.169.254.169.254.1time.repeat.rebind.network/'
+    try:
+        eq(sanitize_user_input_url(url), ipaddress.IPv4Address('8.8.8.8'))  # this is one possible output
+    except PossibleSSRFAttempt as e:
+        eq(e.reason, 'is_link_local')
```

Comment:
Is it expected that this except block will run every time this test is run? If yes, should probably assert that
```suggestion
        assert False, "PossibleSSRFAttempt not raised"
    except PossibleSSRFAttempt as e:
        eq(e.reason, 'is_link_local')
```
Aside: I think I don't understand what this is testing.

Reply:
It's actually a very weird test. The domain resolves non-deterministically to either 8.8.8.8 (which is innocuous) OR a link-local IP (which would be a possible SSRF attempt). So this test just kind of checks that it resolves to one of those two things but doesn't care which one.

I'm not exactly sure what extra utility we get from this example versus one that just deterministically resolves to a link-local address, which we also have; I just took the list of examples from an article about SSRF and tried to cover each example. I think I'd advocate to keep this test as a reminder to in the near future handle rebindings (where a DNS entry changes what IP it resolves to repeatedly and quickly) more intentionally